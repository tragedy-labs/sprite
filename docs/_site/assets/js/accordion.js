/*
 *   This content is licensed according to the W3C Software License at
 *   https://www.w3.org/copyright/software-license-2023/
 *
 *   Simple accordion pattern example
 */

'use strict';

class Accordion {
  constructor(domNode) {
    this.root = domNode;
    this.button = this.root.querySelector('button[aria-expanded]');
    this.icon = this.root.querySelector('svg');
    const controlsId = this.button.getAttribute('aria-controls');
    this.content = document.getElementById(controlsId);

    this.open = this.button.getAttribute('aria-expanded') === 'true';

    // add event listeners
    this.button.addEventListener('click', this.onButtonClick.bind(this));
    this.toggle(false);
  }

  onButtonClick() {
    this.toggle(!this.open);
  }

  toggle(open) {
    // don't do anything if the open state doesn't change
    if (open === this.open) {
      return;
    }

    // update the internal state
    this.open = open;

    // handle DOM updates
    this.button.setAttribute('aria-expanded', `${open}`);
    if (open) {
      this.content.removeAttribute('hidden');
      this.icon.classList.add('open');
    } else {
      this.content.setAttribute('hidden', '');
      this.icon.classList.remove('open');
    }
  }

  // Add public open and close methods for convenience
  open() {
    this.toggle(true);
  }

  close() {
    this.toggle(false);
  }
}

const initializeAccordions = () => {
  // init accordions
  const accordions = document.querySelectorAll('.menuItem');
  accordions.forEach((accordionEl) => {
    new Accordion(accordionEl);
  });
};

document.addEventListener('DOMContentLoaded', initializeAccordions);
