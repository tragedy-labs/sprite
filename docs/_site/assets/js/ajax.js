// Constants
const CONTENT_CONTAINER_ID = "content";
const PARSER = new DOMParser();

// State
let historyStack = [];
let currentStateIndex = -1;

// Functions
async function fetchDocument(url) {
  try {
    const response = await fetch(url);
    return await response.text();
  } catch (error) {
    throw new Error(`Failed to fetch document: ${url}`, { cause: error });
  }
}

async function parseHtml(html) {
  return PARSER.parseFromString(html, "text/html");
}

async function updateContent(htmlDoc) {
  const title = htmlDoc.querySelector("title").textContent;
  const contentDiv = htmlDoc.getElementById(CONTENT_CONTAINER_ID);
  const target = document.getElementById(CONTENT_CONTAINER_ID);
  target.innerHTML = contentDiv.innerHTML;
  target.scrollTop = 0;
  document.title = title;

  // Reinitialize link click events for newly loaded content
  createLinkClickEvents(target);
}

function pushToHistory(url) {
  historyStack = historyStack.slice(0, currentStateIndex + 1);
  historyStack.push(url);
  currentStateIndex++;
  history.pushState({ index: currentStateIndex }, "", url);
}

async function navigate(url, addToHistory = true) {
  const urlObj = new URL(url, window.location.origin);

  if (urlObj)
    if (urlObj.hash && urlObj.pathname === window.location.pathname) {
      // Handle local page anchor navigation
      const anchor = document.querySelector(urlObj.hash);
      if (anchor) {
        anchor.scrollIntoView();
        addToHistory && pushToHistory(url);
        return;
      }
    }

  // Handle Local Navigation
  const html = await fetchDocument(url);
  const htmlDoc = await parseHtml(html);
  await updateContent(htmlDoc);

  addToHistory && pushToHistory(url);
}

function createLinkClickEvents(root) {
  const links = root.querySelectorAll("a");
  links.forEach((link) => {
    link.addEventListener("click", (e) => {
      const url = e.currentTarget.href;
      if (isLocalLink(url)) {
        e.preventDefault();
        navigate(url);
      }
    });
  });
}

function isLocalLink(url) {
  const linkUrl = new URL(url, window.location.origin);
  return linkUrl.origin === window.location.origin;
}

window.addEventListener("popstate", (e) => {
  if (e.state && e.state.index !== undefined) {
    currentStateIndex = e.state.index;
    navigate(historyStack[currentStateIndex], false);
  }
});

// Initialize
function initAjax() {
  createLinkClickEvents(document);
  // Initialize the first state
  historyStack.push(window.location.href);
  currentStateIndex++;
  history.replaceState({ index: currentStateIndex }, "", window.location.href);
}

// Call init
document.addEventListener("DOMContentLoaded", initAjax);
