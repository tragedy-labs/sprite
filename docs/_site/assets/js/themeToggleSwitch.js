function initializeThemeToggle() {
  const element = document.getElementById("theme_toggle");
  const theme = localStorage.getItem("theme");

  const prefersTheme = (theme) =>
    window.matchMedia(`(prefers-color-scheme: ${theme})`).matches;

  const setTheme = (theme) => {
    localStorage.setItem("theme", theme);
    document.body.className = theme;
    setSwitchState(theme);
  };

  const setSwitchState = (theme) => {
    switch (theme) {
      case "dark":
        element.checked = true;
        break;
      case "light":
        element.checked = false;
        break;
      default:
        throw new Error("unknown theme");
    }
  };

  const determinePosition = () => {
    if (theme) {
      setTheme(theme);
    } else {
      if (prefersTheme("dark")) {
        setTheme("dark");
        return;
      } else {
        setTheme("light");
      }
    }
  };

  element.addEventListener("click", () => {
    if (element.checked) {
      setTheme("dark");
    } else {
      setTheme("light");
    }
  });

  determinePosition();
}

document.addEventListener("DOMContentLoaded", initializeThemeToggle);
