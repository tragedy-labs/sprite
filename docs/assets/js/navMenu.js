document.addEventListener("DOMContentLoaded", function () {
  const content = document.getElementById("content");
  const mainNav = document.getElementById("main_nav");
  const navButton = document.getElementById("nav_button");
  const navIndicator = navButton.querySelector("svg");
  const links = mainNav.querySelectorAll("a");

  const closeIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="21"></line>
    <line x1="21" y1="4" x2="3" y2="21"></line>
  `;

  const openIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="4"></line>
    <line x1="3" y1="12" x2="21" y2="12"></line>
    <line x1="3" y1="20" x2="21" y2="20"></line>
  `;

  let isOpen = false;
  let currentTranslation = 0;
  let previousTouchPosition = null;

  function handleTouchEnd() {
    previousTouchPosition = null;

    if (currentTranslation < -80) {
      closeMenu();
    }
    currentTranslation = 0;
    mainNav.style.transform = null;
  }

  function handleTouchStart(event) {
    previousTouchPosition = event.touches[0].clientX;
  }

  function handleTouchMove(event) {
    const currentTouchPosition = event.touches[0].clientX;
    const delta = currentTouchPosition - previousTouchPosition;
    previousTouchPosition = currentTouchPosition;

    currentTranslation += delta;

    if (currentTranslation > 0) {
      currentTranslation = 0;
    }

    mainNav.style.transform = `translate3d(${currentTranslation}px, 0, 0)`;

    const mainNavWidth = mainNav.offsetWidth;
    const ratio = (mainNavWidth + currentTranslation) / mainNavWidth;

    content.style.opacity = 0.65 * (1 - ratio) + 0.35;
  }

  function escHandler(e) {
    if (e.key === "Escape") {
      closeMenu();
    }
  }

  function enableTouchEvents() {
    mainNav.addEventListener("touchstart", handleTouchStart);
    mainNav.addEventListener("touchmove", handleTouchMove);
    mainNav.addEventListener("touchend", handleTouchEnd);
    content.addEventListener("touchstart", closeMenu);
    content.addEventListener("mousedown", closeMenu);
    document.addEventListener("keydown", escHandler);
    // TODO: This could be handled in the ajax script
    // since it already handles link clicks
    links.forEach((link) => {
      link.addEventListener("click", closeMenu);
    });
  }

  function disableTouchEvents() {
    mainNav.removeEventListener("touchstart", handleTouchStart);
    mainNav.removeEventListener("touchmove", handleTouchMove);
    mainNav.removeEventListener("touchend", handleTouchEnd);
    content.removeEventListener("touchstart", closeMenu);
    content.removeEventListener("mousedown", closeMenu);
    document.removeEventListener("keydown", escHandler);
    links.forEach((link) => {
      link.removeEventListener("click", closeMenu);
    });
  }

  function closeMenu() {
    toggleNavState();
    disableTouchEvents();
    navIndicator.innerHTML = openIconMarkup;
    content.style.opacity = null;
  }

  function openMenu() {
    toggleNavState();
    enableTouchEvents();
    navIndicator.innerHTML = closeIconMarkup;
  }

  function toggleNavState() {
    content.classList.toggle("fade");
    mainNav.classList.toggle("open");
    isOpen = !isOpen;
  }

  function handleNavButtonClick() {
    if (isOpen) {
      closeMenu();
    } else {
      openMenu();
    }
  }

  navButton.addEventListener("click", handleNavButtonClick);
});
