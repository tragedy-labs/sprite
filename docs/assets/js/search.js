import data from "/assets/js/searchData.js";
import serializedIndex from "./assets/js/serialized.js";
//import lunr from "lunr"; // assuming lunr is a separate module

// Constants
const MIN_SEARCH_LENGTH = 3;

// State definition
const state = {
  filters: {
    class: true,
    method: true,
    type: true,
  },
  results: {
    filtered: [],
    full: [],
  },
  visible: false,
};

// Elements for reference
const elements = {
  input: null,
  resultsBox: null,
  resultList: null,
  numberOfResults: null,
  numberResultsShowing: null,
  backdrop: null,
  filters: null,
  searchLabel: null,
  close: null,
};

// Initialize search
async function initializeSearch() {
  const root = document.getElementById("search_button");
  elements.input = document.getElementById("searchInput");
  elements.resultsBox = document.getElementById("searchResults");
  elements.resultList = document.getElementById("resultList");
  elements.numberOfResults = document.getElementById("numberOfResults");
  elements.numberResultsShowing = document.getElementById("numberResultsShowing");
  elements.filters = document.getElementById("searchFilters")
    .getElementsByTagName("input");
  elements.searchLabel = document.getElementById("searchLabel");
  elements.close = elements.resultsBox.querySelector("#close_search_dialog");

  // Add event listeners
  elements.input.addEventListener("input", search);
  elements.close.addEventListener("click", hideSearchBox);

  const openSearch = document.getElementById('search_button');

  openSearch.addEventListener('click', showResultsBox);
  document.addEventListener('keydown', (e)=>{
    if (e.key === '/') {
      e.preventDefault();
      showResultsBox();
    };
  }, false);

  // Load the index
  const index = lunr.Index.load(serializedIndex);

  // Initialize search filter toggles
  for (let i = 0; i < elements.filters.length; ++i) {
    elements.filters[i].addEventListener("change", () => {
      state.filters[elements.filters[i].value] = elements.filters[i].checked;
      updateResultList();
    });
  }

  // Define functions
  function appendBackdrop() {
    elements.backdrop = document.createElement("div");
    elements.backdrop.id = "backdrop";
    elements.backdrop.addEventListener("click", hideSearchBox);
    document.body.appendChild(elements.backdrop);
  }

  function removeBackdrop() {
    elements.backdrop.removeEventListener("click", appendBackdrop);
    elements.backdrop.remove();
    elements.backdrop = undefined;
  }

  function handleKeyboardSelection() {
    const items = elements.list.querySelectorAll("a");
    if (items.length > 0) {
      if (document.activeElement === elements.input) {
        e.preventDefault();
        items[0].focus();
      } else {
        for (let i = 0; i < results.length; i++) {
          if (items[i] === document.activeElement) {
            e.preventDefault();
            if (i + 1 < items.length) {
              items[i + 1].focus();
            } else {
              elements.input.focus();
            }
            break;
          }
        }
      }
    }
  }

  function checkKeyPress(e) {
    if (e.key === "Tab") {
      handleKeyboardSelection(e);
    }
    if (e.key === "Escape") {
      hideSearchBox();
    }
  }

  function showResultsBox() {
    appendBackdrop();
    state.visible = true;
    elements.resultsBox.classList.remove("hidden");
    elements.input.focus();
    document.addEventListener("keydown", checkKeyPress);
  }

  function hideSearchBox() {
    state.visible = false;
    elements.resultsBox.classList.add("hidden");
    document.removeEventListener("keydown", checkKeyPress);
    removeBackdrop();
  }

  function updateResultList() {
    state.results.filtered = [];
    elements.resultList.innerHTML = "";
    // elements.numberOfResults.innerHTML = state.results.full.length;
    // elements.numberResultsShowing.innerHTML = state.results.filtered.length;

    if (state.results.full.length > 0) {
      state.results.full.forEach((result) => {
        const doc = data.find((doc) => doc.name === result.ref);
        const type = doc.type === "interface" ? "type" : doc.type;
        if (state.filters[type]) {
          state.results.filtered.push(result.ref);
          const section = document.createElement("section");
          const desc = document.createElement("p");
          const link = document.createElement("a");
          const li = document.createElement("li");
          li.classList.add(doc.type);

          link.href = `/sprite/doc/${doc.name}`;
          link.className = "searchResultLink";
          link.addEventListener("click", () => {});
          link.textContent = doc.parent
            ? `${doc.parent}.${doc.name}`
            : doc.name;

          desc.textContent = doc.desc;

          section.appendChild(link);
          section.appendChild(desc);

          li.appendChild(section);

          elements.resultList.appendChild(li);
        }
      });

      //elements.numberResultsShowing.innerHTML = state.results.filtered.length;
    }
  }

  function displayResults() {
    if (!state.visible) {
      showResultsBox();
    }
    updateResultList();
  }

  function search() {
    if (elements.input.value.length > 0) {
      elements.searchLabel.classList.add("hidden");
    } else {
      elements.searchLabel.classList.remove("hidden");
    }

    if (elements.input.value.length >= MIN_SEARCH_LENGTH) {
      const initial = index.search(elements.input.value);

      if (initial.length === 0) {
        const wildcardResults = index.search(`${elements.input.value}*`);
        state.results.full =
          wildcardResults.length > 0
            ? wildcardResults
            : index.search(`*${elements.input.value}*`);
      } else {
        state.results.full = initial;
      }

      displayResults();
    }
  }
}

document.addEventListener("DOMContentLoaded", initializeSearch);