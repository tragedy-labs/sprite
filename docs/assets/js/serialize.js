const data = require("./searchData.js");
const lunr = require("./lunr.js");

const fs = require("fs");
const path = require("path");

const config = {
  minSearchLength: 3,
  indexedFields: ["name", "desc", "type", "parent"],
};

// Build the index using Lunr.js
const index = lunr(function () {
  this.ref("name");

  config.indexedFields.forEach((field) => {
    this.field(field);
  }, this);

  data.forEach((doc) => {
    this.add(doc);
  }, this);
});

fs.writeFileSync(
  path.join(__dirname, "./serialized.json"),
  JSON.stringify(index, null, 2)
);
