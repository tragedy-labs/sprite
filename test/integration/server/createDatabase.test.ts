// Lib
import { variables } from '@test/variables.js';
import { testClient } from './testClient.js';

const newDatabaseName = 'CreateDatabaseTestDatabase';

describe('SpriteServer.createDatabase', () => {
  afterAll(async () => {
    await testClient.dropDatabase(newDatabaseName);
  });

  it('should create a database', async () => {
    /* Arrange */
    await testClient.createDatabase(newDatabaseName);

    /* Act */
    const list = await testClient.listDatabases();

    /* Assert */
    const databaseCreated = list.includes(newDatabaseName);
    expect(databaseCreated).toBe(true);
  });

  it('should propagate errors from the database', async () => {
    /* Arrange & Act */
    const createPromise = testClient.createDatabase(newDatabaseName);

    /* Assert */
    // Expect an error because the database already exists
    await expect(createPromise).rejects.toMatchSnapshot();
  });
});
