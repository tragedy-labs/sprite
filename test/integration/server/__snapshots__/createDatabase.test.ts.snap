// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpriteServer.createDatabase should propagate errors from the database 1`] = `[Error: Failed to create database "CreateDatabaseTestDatabase".]`;
